lib/open_api_parser/reference.rb

```diff
@@ -0,0 +1,108 @@
+module OpenApiParser
+  # Responsible for interpreting a $ref value and
+  # resolving it to a raw specification given a base URI.
+  class Reference
+    # The resolved document. This gets set only after calling `#resolve`.
+    attr_reader :referrent_document
+
+    # Pointer of the referrent_document if it's embedded in a larger document.
+    # This gets set only after calling `#resolve`.
+    # Empty string means the whole document.
+    attr_reader :referrent_pointer
+
+    def initialize(raw_uri)
+      @raw_uri = raw_uri
+      @resolved = false
+    end
+
+    # Sets referrent_document and referrent_pointer to the resolved
+    # raw specification and pointer, respectively.
+    #
+    # @param base_path [String] Location of the document where the $ref originates.
+    # @param base_pointer [String] Location of the $ref within the document.
+    # @param current_document [Hash] Document where the $ref originates.
+    # @param file_cache [OpenApiParser::FileCache] File cache instance.
+    # @return [Boolean] Whether the referrent has been fully expanded.
+    def resolve(base_path, base_pointer, current_document, file_cache)
+      if @resolved
+        fail 'Do not try to resolve an already resolved reference.'
+      end
+      @resolved = true
```

Comment:
Can we memoize this method such that subsequent calls return the same value rather than failing?

Reply:
I've tried reducing the number of states that the `Reference` object holds: 134f3f814c7cc9a61e8dfa109c21c680bba9b63b so that `#resolve` can be called any number of times. This way, a `Reference` is a representation of a `{"$ref": ".."}` object, which can be resolved in many different contexts.

If you had performance concerns, I can look into caching the reference resolution result at the Document level.